src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala

```diff
@@ -72,78 +74,87 @@ trait NihService extends LazyLogging {
   }
 
-  private def getWhitelist(): Set[String] = {
-    val (bucket, file) = (FireCloudConfig.Nih.whitelistBucket, FireCloudConfig.Nih.whitelistFile)
-    val whitelist = Source.fromInputStream(googleDao.getBucketObjectAsInputStream(bucket, file))
+  // This syncs all of the whitelists in full
+  def syncNihWhitelistsFull: Future[PerRequestMessage] = {
+    val whitelistSyncResults = Future.sequence(nihWhitelists.map(syncNihWhitelistAllUsers))
 
-    whitelist.getLines().toSet
+    whitelistSyncResults map { response =>
+      if(response.forall(x => x)) RequestComplete(NoContent)
+      else RequestCompleteWithErrorReport(InternalServerError, "Error synchronizing NIH whitelist")
+    }
   }
 
-  // This syncs the entire whitelist
-  def syncWhitelistFull(): Future[PerRequestMessage] = {
-    val whitelist = getWhitelist()
+  // This syncs the specified whitelist in full
+  def syncNihWhitelistAllUsers(nihWhitelist: NihWhitelist): Future[Boolean] = {
+    val whitelistUsers = downloadNihWhitelist(nihWhitelist)
 
     // The list of users that, according to Thurloe, have active links and are
-    // on the dbGap whitelist
+    // on the specified whitelist
     val nihEnabledFireCloudUsers = getCurrentNihUsernameMap(thurloeDao) map { mapping =>
-      mapping.collect { case (fcUser, nihUser) if whitelist contains nihUser => fcUser }.toSeq
+      mapping.collect { case (fcUser, nihUser) if whitelistUsers contains nihUser => fcUser }.toSeq
     }
 
     val memberList = nihEnabledFireCloudUsers map { subjectIds => {
       RawlsGroupMemberList(userSubjectIds = Some(subjectIds))
     }}
 
-    // The request to rawls to completely overwrite the dbGapAuthorizedUsers group
+    // The request to rawls to completely overwrite the group
     // with the list of actively linked users on the whitelist
-    val rawlsRequest = memberList flatMap { members =>
-      rawlsDao.adminOverwriteGroupMembership(FireCloudConfig.Nih.rawlsGroupName, members)
+    memberList flatMap { members =>
+      rawlsDao.adminOverwriteGroupMembership(nihWhitelist.groupToSync, members)
     }
+  }
 
-    rawlsRequest map { response =>
-      if(response) RequestComplete(NoContent)
-      else RequestCompleteWithErrorReport(InternalServerError, "Error synchronizing NIH whitelist.")
+  private def linkNihAccount(userInfo: UserInfo, nihLink: NihLink): Future[Try[Unit]] = {
+    val profilePropertyMap = nihLink.propertyValueMap
+
+    thurloeDao.saveKeyValues(userInfo, profilePropertyMap)
+  }
+
+  def updateNihLinkAndSyncSelf(userInfo: UserInfo, nihLink: NihLink): Future[PerRequestMessage] = {
+    //first, link the NIH account. it doesn't matter if they're on a whitelist or not
+    //next, sync the appropriate rawls groups based on which whitelists the user is a member of
+    val linkResult = for {
+      linkNihResult <- linkNihAccount(userInfo, nihLink)
+      _ <- Future.sequence(nihWhitelists.map(syncNihWhitelistForUser(userInfo.getUniqueId, nihLink.linkedNihUsername, _)))
```

Comment:
I think of syncing the various whitelists as side effects of NIH-linking. The result of the overall operation doesn't depend on what happens with the whitelists- so I think fire and forget is OK here.